File: ios/Classes/EdgeGenaiPlugin.swift
import Flutter
import UIKit
#if canImport(FoundationModels)
  import FoundationModels
#endif

public class EdgeGenaiPlugin: NSObject, FlutterPlugin, EdgeGenaiHostApi {
  private var pendingPrompt: String?
  private var pendingOptions: EdgeGenaiGenerationOptions?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = EdgeGenaiPlugin()
    EdgeGenaiHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
    DownloadProgressStreamHandler.register(
      with: registrar.messenger(), streamHandler: EdgeGenaiDownloadProgressStreamHandler())
    GenerateContentChunkStreamHandler.register(
      with: registrar.messenger(),
      streamHandler: EdgeGenaiGenerateContentStreamHandler { [weak instance] in
        guard let prompt = instance?.pendingPrompt else { return nil }
        return (prompt, instance?.pendingOptions)
      })
  }

  func checkAvailability(completion: @escaping (Result<EdgeGenaiAvailability, Error>) -> Void) {
    #if canImport(FoundationModels)
      if #available(iOS 26.0, *) {
        switch SystemLanguageModel.default.availability {
        case .available:
          completion(.success(.available))
          return
        case .unavailable(.deviceNotEligible):
          completion(.success(.unavailable))
          return
        case .unavailable(.appleIntelligenceNotEnabled):
          completion(.success(.notYetReady))
          return
        case .unavailable(.modelNotReady):
          completion(.success(.downloadable))
          return
        case .unavailable:
          completion(.success(.unavailable))
          return
        }
      }
    #endif
    completion(.success(.unavailable))
  }

  func startGenerateContent(prompt: String, options: EdgeGenaiGenerationOptions?) throws {
    pendingPrompt = prompt
    pendingOptions = options
  }
}

/// There's nothing for the app to download on iOS — Apple Intelligence is
/// enabled system-wide in Settings — so this immediately reports completion.
private class EdgeGenaiDownloadProgressStreamHandler: DownloadProgressStreamHandler {
  override func onListen(
    withArguments arguments: Any?, sink: PigeonEventSink<EdgeGenaiDownloadProgress>
  ) {
    sink.success(EdgeGenaiDownloadProgress(status: .completed, bytesDownloaded: nil))
    sink.endOfStream()
  }
}

/// Starts generation for the prompt/options stashed via `startGenerateContent` when Flutter
/// starts listening, and streams the cumulative response text as it's generated.
private class EdgeGenaiGenerateContentStreamHandler: GenerateContentChunkStreamHandler {
  private let takePendingRequest: () -> (String, EdgeGenaiGenerationOptions?)?

  init(takePendingRequest: @escaping () -> (String, EdgeGenaiGenerationOptions?)?) {
    self.takePendingRequest = takePendingRequest
    super.init()
  }

  override func onListen(withArguments arguments: Any?, sink: PigeonEventSink<String>) {
    guard let (prompt, options) = takePendingRequest() else {
      sink.error(
        code: "no_prompt", message: "startGenerateContent must be called before listening.",
        details: nil)
      return
    }
    #if canImport(FoundationModels)
      if #available(iOS 26.0, *) {
        Task {
          do {
            let generationOptions = GenerationOptions(
              temperature: options?.temperature,
              maximumResponseTokens: options?.maxOutputTokens.map { Int($0) })
            let session = LanguageModelSession()
            let stream = session.streamResponse(to: prompt, options: generationOptions)
            for try await snapshot in stream {
              sink.success(snapshot.content)
            }
            sink.endOfStream()
          } catch {
            sink.error(code: "generate_content_failed", message: error.localizedDescription, details: nil)
          }
        }
        return
      }
    #endif
    sink.error(
      code: "unavailable", message: "The on-device model isn't available.", details: nil)
  }
}

File: android/src/main/kotlin/com/example/edge_genai/EdgeGenaiPlugin.kt
package com.example.edge_genai

import com.google.mlkit.genai.common.DownloadStatus
import com.google.mlkit.genai.common.FeatureStatus
import com.google.mlkit.genai.prompt.Generation
import com.google.mlkit.genai.prompt.GenerativeModel
import io.flutter.embedding.engine.plugins.FlutterPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/** EdgeGenaiPlugin */
class EdgeGenaiPlugin :
    FlutterPlugin,
    EdgeGenaiHostApi {
    private var pluginBinding: FlutterPlugin.FlutterPluginBinding? = null
    private val scope = CoroutineScope(Dispatchers.Main)
    private val generativeModel by lazy { Generation.getClient() }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        pluginBinding = flutterPluginBinding
        EdgeGenaiHostApi.setUp(flutterPluginBinding.binaryMessenger, this)
        DownloadProgressStreamHandler.register(
            flutterPluginBinding.binaryMessenger,
            EdgeGenaiDownloadProgressStreamHandler(scope, generativeModel),
        )
    }

    override fun checkAvailability(callback: (Result<EdgeGenaiAvailability>) -> Unit) {
        scope.launch {
            val availability =
                try {
                    when (generativeModel.checkStatus()) {
                        FeatureStatus.AVAILABLE -> EdgeGenaiAvailability.AVAILABLE
                        FeatureStatus.DOWNLOADABLE, FeatureStatus.DOWNLOADING ->
                            EdgeGenaiAvailability.DOWNLOADABLE
                        else -> EdgeGenaiAvailability.UNAVAILABLE
                    }
                } catch (e: Exception) {
                    callback(Result.failure(e))
                    return@launch
                }
            callback(Result.success(availability))
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        pluginBinding?.let { EdgeGenaiHostApi.setUp(it.binaryMessenger, null) }
        pluginBinding = null
    }

    override fun generateContent(
        prompt: String,
        callback: (Result<String>) -> Unit
    ) {
        scope.launch {
            try {
                val response = generativeModel.generateContent(prompt)
                callback(Result.success(response.candidates.first().text))
            } catch (e: Exception) {
                callback(Result.failure(e))
            }
        }
    }
}

/** Triggers the Gemini Nano download when Flutter starts listening, and streams its progress. */
private class EdgeGenaiDownloadProgressStreamHandler(
    private val scope: CoroutineScope,
    private val generativeModel: GenerativeModel
) : DownloadProgressStreamHandler() {
    override fun onListen(
        p0: Any?,
        sink: PigeonEventSink<EdgeGenaiDownloadProgress>
    ) {
        scope.launch {
            generativeModel.download().collect { status ->
                when (status) {
                    is DownloadStatus.DownloadStarted ->
                        sink.success(
                            EdgeGenaiDownloadProgress(EdgeGenaiDownloadStatus.STARTED, null)
                        )
                    is DownloadStatus.DownloadProgress ->
                        sink.success(
                            EdgeGenaiDownloadProgress(
                                EdgeGenaiDownloadStatus.IN_PROGRESS,
                                status.totalBytesDownloaded
                            )
                        )
                    is DownloadStatus.DownloadCompleted -> {
                        sink.success(
                            EdgeGenaiDownloadProgress(EdgeGenaiDownloadStatus.COMPLETED, null)
                        )
                        sink.endOfStream()
                    }
                    is DownloadStatus.DownloadFailed ->
                        sink.error("download_failed", status.e.message, null)
                }
            }
        }
    }
}
